CreateProcess error=206, The filename or extension is too long

Using 0.9.1, I received the following error on trying to build the native image:

[ERROR] Failed to execute goal org.graalvm.buildtools:native-maven-plugin:0.9.1:build (build-native) on project SBProject: Building image with C:\graalvm\bin\native-image.cmd failed: Cannot run program "C:\graalvm\bin\native-image.cmd" (in directory "D:\projects\SBProject\target"): CreateProcess error=206, The filename or extension is too long

@lazar-mitrovic 
`
> Task :build
Skipping task ':build' as it has no actions.
:build (Thread[Execution worker for ':' Thread 4,5,main]) completed. Took 0.001 secs.
:nativeBuild (Thread[Execution worker for ':' Thread 4,5,main]) started.

> Task :nativeBuild FAILED
Caching disabled for task ':nativeBuild' because:
  Build cache is disabled
Task ':nativeBuild' is not up-to-date because:
  Task has failed previously.
Starting process 'command 'D:\graalvm\bin\native-image.cmd''. Working directory: D:\shileep-ucenter\build Command: D:\graalvm\bin\native-image.cmd -cp D:\shileep-ucenter\build\classes\
java\main;D:\shileep-ucenter\build\classes\kotlin\main;D:\shileep-ucenter\build\resources\main;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.experimental\spring
-native\0.10.1\3ce5112025340bf53be27af1f8e2909b25142d6b\spring-native-0.10.1.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.graalvm.buildtools\junit-platform-native\0.9.1\aa
03ea73b77a00b729e61e686931e8032f2f3d91\junit-platform-native-0.9.1.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-boot-starter-data-jpa\2.5.2\503
8e48790990bc137c9a7e1cd97901a1350eb37\spring-boot-starter-data-jpa-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-boot-starter-web\2.5.2\9d
68318f6607728e253030aa145e529bae9b0883\spring-boot-starter-web-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-boot-starter-json\2.5.2\6af2e
2d4fcf02fcda700dbbcd4643e78d31bcc8c\spring-boot-starter-json-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml.jackson.datatype\jackson-datatype-jdk8\2.12.3\774
24ea087313312e308dae5ff8445608aabb5e1\jackson-datatype-jdk8-2.12.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml.jackson.datatype\jackson-datatype-jsr310\2.12.3\f
69c636438dcf19c49960c1fe8901320ab85f989\jackson-datatype-jsr310-2.12.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml.jackson.module\jackson-module-parameter-names
\2.12.3\592a882beaf1bd57b8fe960b937a2706b090b4d7\jackson-module-parameter-names-2.12.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml.jackson.core\jackson-databind
\2.12.3\d6153f8fc60c479ab0f9efb35c034526436a4953\jackson-databind-2.12.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml.jackson.core\jackson-annotations\2.12.3\727
5513412694a1aafd08c0287f48469fa0e6e17\jackson-annotations-2.12.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml.jackson.core\jackson-core\2.12.3\deb23fe2a7f2b773e1
8ced2b50d4acc1df8fa366\jackson-core-2.12.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml.jackson.module\jackson-module-kotlin\2.12.3\c8da7d998ed2e9253e62fdf15d111
2c54b3ce6db\jackson-module-kotlin-2.12.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-reflect\1.5.20\63810951f09a27f7827977f1bc21094042a915e\kotlin
-reflect-1.5.20.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-jdk8\1.5.20\636c5653641cd956de9aee5792155b07ea49825e\kotlin-stdlib-jdk8-1.5.20.
jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\mysql\mysql-connector-java\8.0.25\f8b9123acd13058c941aff25f308c9ed8000bb73\mysql-connector-java-8.0.25.jar;C:\Users\12444\.gradle\
caches\modules-2\files-2.1\org.junit.platform\junit-platform-reporting\1.7.2\8c6521bce2744fbce1c12a2088e05e754911213\junit-platform-reporting-1.7.2.jar;C:\Users\12444\.gradle\caches\mo
dules-2\files-2.1\org.junit.platform\junit-platform-launcher\1.7.2\9bcb72e95d291c9ff45c3f8c0a054abc764b4205\junit-platform-launcher-1.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\fi
les-2.1\org.junit.jupiter\junit-jupiter\5.7.2\62faa742964a9d8dab8fdb4a0eab7b01441c171f\junit-jupiter-5.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.junit.jupiter\junit
-jupiter-engine\5.7.2\9415680a889f00b8205a094c5c487bc69dc7077d\junit-jupiter-engine-5.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.junit.platform\junit-platform-engine
\1.7.2\2573770b46b8a199ed5f6b0f96fb99e468bfe891\junit-platform-engine-1.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.junit.jupiter\junit-jupiter-params\5.7.2\685f832f8
c54dd40100f646d61aca88ed9545421\junit-jupiter-params-5.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.junit.jupiter\junit-jupiter-api\5.7.2\f4b4079732a9c537983324cfa4e46
655f21d2c56\junit-jupiter-api-5.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.junit.platform\junit-platform-commons\1.7.2\34adfea6c13fc4a996cf38cdad80800ce850d198\junit
-platform-commons-1.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.junit.platform\junit-platform-console\1.7.2\f98888d66074f6c060054675ab8366885f077143\junit-platform-co
nsole-1.7.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-boot-starter-aop\2.5.2\1877448bb875154c2558e385f4590c648ada5f11\spring-boot-starter-ao
p-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-boot-starter-jdbc\2.5.2\de9142ad6a67d2ec7d58465ab8d2fcf759f24d25\spring-boot-starter-jdbc-
2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\jakarta.transaction\jakarta.transaction-api\1.3.3\c4179d48720a1e87202115fbed6089bdc4195405\jakarta.transaction-api-1.3.3.jar
;C:\Users\12444\.gradle\caches\modules-2\files-2.1\jakarta.persistence\jakarta.persistence-api\2.2.3\8f6ea5daedc614f07a3654a455660145286f024e\jakarta.persistence-api-2.2.3.jar;C:\Users
\12444\.gradle\caches\modules-2\files-2.1\org.hibernate\hibernate-core\5.4.32.Final\99a5e10bf455337014c190e141ec631e9ff71663\hibernate-core-5.4.32.Final.jar;C:\Users\12444\.gradle\cach
es\modules-2\files-2.1\org.springframework.data\spring-data-jpa\2.5.2\37a27171338da87485cdf66a40596195433d1db\spring-data-jpa-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.
1\org.springframework\spring-aspects\5.3.8\863f531cc3ea9d82d391d17c8eb33732e7a864b6\spring-aspects-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\
spring-boot-starter\2.5.2\5aa5345f60cd3fbca331c05015e0700553448d66\spring-boot-starter-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-boot-
starter-tomcat\2.5.2\ed7d2ef9efef6145bdc8c9c4dd1790f51673f27d\spring-boot-starter-tomcat-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework\spring-webmvc\5
.3.8\2cfa5e12bf103e2c82db78ce198e455dab456465\spring-webmvc-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework\spring-web\5.3.8\daa288e67b0f2e09a033500d5ce
8406677c5045c\spring-web-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-jdk7\1.5.20\218b60e1d446d1e0a18bc7aa8663634b136fbcc5\kotlin-stdl
ib-jdk7-1.5.20.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib\1.5.20\9de35cc611bcecec8edce1d56d8e659953806751\kotlin-stdlib-1.5.20.jar;C:\User
s\12444\.gradle\caches\modules-2\files-2.1\org.apiguardian\apiguardian-api\1.1.0\fc9dff4bb36d627bdc553de77e1f17efd790876c\apiguardian-api-1.1.0.jar;C:\Users\12444\.gradle\caches\module
s-2\files-2.1\org.springframework.boot\spring-boot-autoconfigure\2.5.2\aa924c5d0d375f4b2b66f211704c24d40a7e647b\spring-boot-autoconfigure-2.5.2.jar;C:\Users\12444\.gradle\caches\module
s-2\files-2.1\org.springframework.boot\spring-boot\2.5.2\bd6dc87b5ad870dbf635d0c7d9a3006666c9e797\spring-boot-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springfram
ework\spring-context\5.3.8\c367a05423e963c222e38a6a88b97d44de3880ca\spring-context-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework\spring-aop\5.3.8\1377
f80f938b1fc7eabe9e6c4f6895e77e3bec40\spring-aop-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.aspectj\aspectjweaver\1.9.6\ee3b73aa16df35179255f17354d9dfd8e7822835\asp
ectjweaver-1.9.6.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.zaxxer\HikariCP\4.0.3\107cbdf0db6780a065f895ae9d8fbf3bb0e1c21f\HikariCP-4.0.3.jar;C:\Users\12444\.gradle\cach
es\modules-2\files-2.1\org.springframework\spring-orm\5.3.8\bac8889d4621f2f47ee2b118ebbddf61dd299239\spring-orm-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springfr
amework\spring-jdbc\5.3.8\cf8b7a2be3565eed94f575ae640e22d9357c7dcd\spring-jdbc-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.hibernate.common\hibernate-commons-annota
tions\5.1.2.Final\e59ffdbc6ad09eeb33507b39ffcf287679a498c8\hibernate-commons-annotations-5.1.2.Final.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.jboss.logging\jboss-loggi
ng\3.4.2.Final\e517b8a93dd9962ed5481345e4d262fdd47c4217\jboss-logging-3.4.2.Final.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.javassist\javassist\3.27.0-GA\f63e6aa899e15e
ca8fdaa402a79af4c417252213\javassist-3.27.0-GA.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\net.bytebuddy\byte-buddy\1.10.22\ef45d7e2cd1c600d279704f492ed5ce2ceb6cdb5\byte-budd
y-1.10.22.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\antlr\antlr\2.7.7\83cd2cd674a217ade95a4bb83a8a14f351f48bd0\antlr-2.7.7.jar;C:\Users\12444\.gradle\caches\modules-2\files
-2.1\org.jboss\jandex\2.2.3.Final\d3865101f0666b63586683bd811d754517f331ab\jandex-2.2.3.Final.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.fasterxml\classmate\1.5.1\3fe0be
d568c62df5e89f4f174c101eab25345b6c\classmate-1.5.1.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.dom4j\dom4j\2.1.3\a75914155a9f5808963170ec20653668a2ffd2fd\dom4j-2.1.3.jar;
C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.glassfish.jaxb\jaxb-runtime\2.3.4\5f3828d3a345ff3d8acb83f07ba04eccf59e15bb\jaxb-runtime-2.3.4.jar;C:\Users\12444\.gradle\caches\mo
dules-2\files-2.1\org.springframework.data\spring-data-commons\2.5.2\25a00968c046736667e086ed9fd254f0e26e7ffc\spring-data-commons-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\file
s-2.1\org.springframework\spring-tx\5.3.8\3202859a7e9560110f1fe8284c28feb009c6f460\spring-tx-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework\spring-bean
s\5.3.8\3d66fed1eebfcd119efcabc6218c813700a21ed\spring-beans-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework\spring-expression\5.3.8\8a14547b76cbae3aeb0
2739e5b38e71835a6bbd8\spring-expression-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework\spring-core\5.3.8\da9b87dacaa5bbf80fad0f7b483988372a00a152\sprin
g-core-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-boot-starter-logging\2.5.2\8b94a1e350f68f8c3a7de6460e5691a84c8c157d\spring-boot-start
er-logging-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\ch.qos.logback\logback-classic\1.2.3\7c4f3c474fb2c041d8028740440937705ebb473a\logback-classic-1.2.3.jar;C:\Users\
12444\.gradle\caches\modules-2\files-2.1\org.apache.logging.log4j\log4j-to-slf4j\2.14.1\ce8a86a3f50a4304749828ce68e7478cafbc8039\log4j-to-slf4j-2.14.1.jar;C:\Users\12444\.gradle\caches
\modules-2\files-2.1\org.slf4j\jul-to-slf4j\1.7.31\f9ff62d83a25a94c1619de06d4015e2797bc849c\jul-to-slf4j-1.7.31.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.slf4j\slf4j-ap
i\1.7.31\9545c9cb71de4c18d97a91e32ef0be6f3f6661b7\slf4j-api-1.7.31.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\jakarta.annotation\jakarta.annotation-api\1.3.5\59eb84ee0d61633
2ff44aba065f3888cf002cd2d\jakarta.annotation-api-1.3.5.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.yaml\snakeyaml\1.28\7cae037c3014350c923776548e71c9feb7a69259\snakeyaml-
1.28.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.apache.tomcat.embed\tomcat-embed-websocket\9.0.48\a34e3133ec3d61f4c2faa30476c4b43cc6385dd6\tomcat-embed-websocket-9.0.48.
jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.apache.tomcat.embed\tomcat-embed-core\9.0.48\f112cd2380d8215e22ac40aff128a1b6daa2f0ac\tomcat-embed-core-9.0.48.jar;C:\Users\12
444\.gradle\caches\modules-2\files-2.1\org.apache.tomcat.embed\tomcat-embed-el\9.0.48\c46a092831493ccf65674cb19235625853ceb3d4\tomcat-embed-el-9.0.48.jar;C:\Users\12444\.gradle\caches\
modules-2\files-2.1\org.jetbrains\annotations\13.0\919f0dfe192fb4e063e7dacadee7f8bb9a2672a9\annotations-13.0.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\
kotlin-stdlib-common\1.5.20\3d79dbd48bf605f4aac1e7028981a1953e245cbb\kotlin-stdlib-common-1.5.20.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.opentest4j\opentest4j\1.2.0\2
8c11eb91f9b6d8e200631d46e20a7f407f2a046\opentest4j-1.2.0.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\jakarta.xml.bind\jakarta.xml.bind-api\2.3.3\48e3b9cfc10752fba3521d6511f41
65bea951801\jakarta.xml.bind-api-2.3.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.glassfish.jaxb\txw2\2.3.4\257fa649d3137a1060d222aefb96b7d1dd5f1286\txw2-2.3.4.jar;C:\Us
ers\12444\.gradle\caches\modules-2\files-2.1\com.sun.istack\istack-commons-runtime\3.0.12\cbbe1a62b0cc6c85972e99d52aaee350153dc530\istack-commons-runtime-3.0.12.jar;C:\Users\12444\.gra
dle\caches\modules-2\files-2.1\com.sun.activation\jakarta.activation\1.2.2\74548703f9851017ce2f556066659438019e7eb5\jakarta.activation-1.2.2.jar;C:\Users\12444\.gradle\caches\modules-2
\files-2.1\org.springframework\spring-jcl\5.3.8\a143c8618eb2be8674c3cf132d9a5c953bb5488\spring-jcl-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\ch.qos.logback\logback-co
re\1.2.3\864344400c3d4d92dfeb0a305dc87d953677c03c\logback-core-1.2.3.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.apache.logging.log4j\log4j-api\2.14.1\cd8858fbbde69f46bce
8db1152c18a43328aae78\log4j-api-2.14.1.jar;D:\shileep-ucenter\build\classes\java\aotTest;D:\shileep-ucenter\build\classes\kotlin\aotTest;D:\shileep-ucenter\build\resources\aotTest;D:\s
hileep-ucenter\build\classes\java\test;D:\shileep-ucenter\build\classes\kotlin\test;D:\shileep-ucenter\build\resources\test;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.spring
framework.boot\spring-boot-starter-test\2.5.2\226e09262316b1ff101797346b2ca56432880353\spring-boot-starter-test-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.mockito\
mockito-junit-jupiter\3.9.0\871745ab6af5a269411ea3c1f99ced82ed079436\mockito-junit-jupiter-3.9.0.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boot\spring-b
oot-test-autoconfigure\2.5.2\223f1311f59eecfd24980f9776bb39e1fef92c0e\spring-boot-test-autoconfigure-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework.boo
t\spring-boot-test\2.5.2\477509ac730e2fb8fe67c0d9f2d0bca21c4316df\spring-boot-test-2.5.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.jayway.jsonpath\json-path\2.5.0\c35ef
29095125b51638d19120f63e2b56eff20e9\json-path-2.5.0.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.assertj\assertj-core\3.19.0\f64cb5690b85e68d5e1e6c6152bfb6e3840a452d\asser
tj-core-3.19.0.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.hamcrest\hamcrest\2.2\1820c0968dba3a11a1b30669bb1f01978a91dedc\hamcrest-2.2.jar;C:\Users\12444\.gradle\caches\m
odules-2\files-2.1\org.mockito\mockito-core\3.9.0\b7573430aea743b26434b44f4f46272af613e660\mockito-core-3.9.0.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.skyscreamer\json
assert\1.5.0\6c9d5fe2f59da598d9aefc1cfc6528ff3cf32df3\jsonassert-1.5.0.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.springframework\spring-test\5.3.8\4ce3ee7725d4ecd13b5c5
2d67bb0a76173f8a34d\spring-test-5.3.8.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.xmlunit\xmlunit-core\2.8.2\8e4e46b87afaaf9b6cfb8de778e473cf7aeb087f\xmlunit-core-2.8.2.j
ar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\net.minidev\json-smart\2.4.7\8d7f4c1530c07c54930935f3da85f48b83b3c109\json-smart-2.4.7.jar;C:\Users\12444\.gradle\caches\modules-2\
files-2.1\jakarta.activation\jakarta.activation-api\1.2.2\99f53adba383cb1bf7c3862844488574b559621f\jakarta.activation-api-1.2.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\ne
t.bytebuddy\byte-buddy-agent\1.10.22\b01df6b71a882b9fde5a608a26e641cd399a4d83\byte-buddy-agent-1.10.22.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.objenesis\objenesis\3.2
\7fadf57620c8b8abdf7519533e5527367cb51f09\objenesis-3.2.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\com.vaadin.external.google\android-json\0.0.20131108.vaadin1\fa26d351fe62a
6a17f5cda1287c1c6110dec413f\android-json-0.0.20131108.vaadin1.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\net.minidev\accessors-smart\2.4.7\3970cfc505e6657ca60f3aa57c849f6043
000d7a\accessors-smart-2.4.7.jar;C:\Users\12444\.gradle\caches\modules-2\files-2.1\org.ow2.asm\asm\9.1\a99500cf6eea30535eeac6be73899d048f8d12a8\asm-9.1.jar --no-fallback -H:Path=native
 -H:Name=shileep-ucenter -H:Class=cn.ziero.shileep.ucenter.ShileepUcenterApplication --no-server
Successfully started process 'command 'D:\graalvm\bin\native-image.cmd''
The command line is too long.
:nativeBuild (Thread[Execution worker for ':' Thread 4,5,main]) completed. Took 0.039 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeBuild'.
> Process 'command 'D:\graalvm\bin\native-image.cmd'' finished with non-zero exit value 1

* Try:
Run with --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':nativeBuild'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:187)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:268)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:185)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:408)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:395)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:388)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:374)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.process.internal.ExecException: Process 'command 'D:\graalvm\bin\native-image.cmd'' finished with non-zero exit value 1
        at org.gradle.process.internal.DefaultExecHandle$ExecResultImpl.assertNormalExitValue(DefaultExecHandle.java:414)
        at org.gradle.process.internal.DefaultExecAction.execute(DefaultExecAction.java:38)
        at org.gradle.api.tasks.AbstractExecTask.exec(AbstractExecTask.java:69)
        at org.graalvm.buildtools.gradle.tasks.NativeBuildTask.exec(NativeBuildTask.java:129)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:29)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$2.run(ExecuteActionsTaskExecuter.java:498)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:483)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:466)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$300(ExecuteActionsTaskExecuter.java:105)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:270)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:248)
        at org.gradle.internal.execution.steps.ExecuteStep.executeInternal(ExecuteStep.java:83)
        at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:37)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:50)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:47)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:37)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:68)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:38)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:50)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:36)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:41)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:74)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:55)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:29)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:54)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:35)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:60)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:27)
        at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:174)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:74)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:45)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:40)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:29)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:36)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:22)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:99)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:92)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:52)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:36)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:84)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:41)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:91)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:49)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:78)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:49)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:105)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:50)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:86)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:86)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:32)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:43)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:31)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.lambda$execute$0(AssignWorkspaceStep.java:40)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution$2.withWorkspace(ExecuteActionsTaskExecuter.java:283)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:40)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:30)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:37)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:27)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:49)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:35)
        at org.gradle.internal.execution.impl.DefaultExecutionEngine$1.execute(DefaultExecutionEngine.java:76)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:184)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:408)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:395)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:388)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:374)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)


* Get more help at https://help.gradle.org

BUILD FAILED in 12s
11 actionable tasks: 5 executed, 6 up-to-date
Watched directory hierarchies: [D:\shileep-ucenter]
`

--- a/docs/src/docs/asciidoc/gradle-plugin.adoc
+++ b/docs/src/docs/asciidoc/gradle-plugin.adoc
@@ -159,6 +159,8 @@ nativeBuild {
 
   // Development options
   agent = true // Enables the reflection agent. Can be also set on command line using '-Pagent'
+
+  useFatJar = true // Instead of passing each jar individually, builds a fat jar
 }
 ```
 
@@ -185,10 +187,54 @@ nativeBuild {
 
   // Development options
   agent.set(true) // Enables the reflection agent. Can be also set on command line using '-Pagent'
+
+
+  useFatJar.set(true) // Instead of passing each jar individually, builds a fat jar
 }
 ```
 
 NOTE: For options that can be set using command-line, if both DSL and command-line options are present, command-line options take precedence.
+
+==== Long classpath and fat jar support
+
+Under Windows, https://github.com/graalvm/native-build-tools/issues/85[it is possible that the length of the classpath exceeds what the operating system supports] when invoking the CLI to build a native image.
+As a consequence, if you are running under Windows, the plugin will automatically shorten the classpath of your project by building a so called "fat jar", which includes all entries from the classpath automatically.
+
+In case this behavior is not required, you can disable the fat jar creation by calling:
+
+.Disabling the fat jar creation
+[role="multi-language-sample"]
+```groovy
+nativeBuild {
+    useFatJar = false
+}
+```
+
+[role="multi-language-sample"]
+```kotlin
+nativeBuild {
+    useFatJar.set(false)
+}
+```
+
+Alternatively, it is possible to use your own fat jar (for example created using the https://imperceptiblethoughts.com/shadow/[Shadow plugin]) by setting the `classpathJar` property directly on the _task_:
+
+.Disabling the fat jar creation
+[role="multi-language-sample"]
+```groovy
+tasks.named("nativeBuild") {
+    classpathJar = myFatJar
+}
+```
+
+[role="multi-language-sample"]
+```kotlin
+tasks.named<BuildNativeImageTask>("nativeBuild") {
+    classpathJar.set(myFatJar)
+}
+```
+
+When the `classpathJar` property is set, the `classpath` property is _ignored_.
 
 [[testing]]
 === Testing support

--- a/docs/src/docs/asciidoc/maven-plugin.adoc
+++ b/docs/src/docs/asciidoc/maven-plugin.adoc
@@ -147,6 +147,24 @@ This step won't be needed as of JUnit 5.8 with a future release of native-maven-
 
 Running `mvn -Pnative -Dskip test` will then build and run native tests.
 
+== Long classpath and shading support
+
+Under Windows, https://github.com/graalvm/native-build-tools/issues/85[it is possible that the length of the classpath exceeds what the operating system supports] when invoking the CLI to build a native image.
+
+If this happens, one option is to use a https://maven.apache.org/plugins/maven-shade-plugin[shaded jar] and use it instead of individual jars on classpath.
+
+First, you'll need to setup the https://maven.apache.org/plugins/maven-shade-plugin[Maven Shade plugin]:
+
+[source,xml,indent=0]
+include::../../../../samples/java-application/pom.xml[tag=shade-plugin]
+
+Then the native plugin needs to be configured to use this jar instead of the full classpath:
+
+[source,xml,indent=0]
+include::../../../../samples/java-application/pom.xml[tag=native-plugin]
+
+Please refer to the https://maven.apache.org/plugins/maven-shade-plugin[Maven Shade plugin documentation] for more details on how to configure shading.
+
 == Javadocs
 
 In addition, you can consult the link:javadocs/native-maven-plugin/index.html[Javadocs of the plugin].

--- a/docs/src/docs/asciidoc/maven-plugin.adoc
+++ b/docs/src/docs/asciidoc/maven-plugin.adoc
@@ -127,6 +127,7 @@ In this case, the arguments that will be passed to the `native-image` executable
 --no-fallback --verbose
 ```
 
+[[testing-support]]
 == JUnit Testing support
 
 In order to use the recommended test listener mode, you need to add following dependency:
@@ -163,7 +164,30 @@ Then the native plugin needs to be configured to use this jar instead of the ful
 [source,xml,indent=0]
 include::../../../../samples/java-application/pom.xml[tag=native-plugin]
 
-Please refer to the https://maven.apache.org/plugins/maven-shade-plugin[Maven Shade plugin documentation] for more details on how to configure shading.
+To be able to <<testing-support,execute tests in native mode>>, you will need more setup:
+
+- Create a `src/assembly/test-jar-with-dependencies.xml` file with the following contents:
+
+[source,xml,indent=0]
+include::../../../../samples/java-application-with-tests/src/assembly/test-jar-with-dependencies.xml[tag=assembly]
+
+- Add the assembly plugin to your `native` profile:
+
+[source,xml,indent=0]
+include::../../../../samples/java-application-with-tests/pom.xml[tag=assembly-plugin]
+
+- Due to a limitation in Maven, you will need to move the tests execution to the "integration-test" phase:
+
+[source,xml,indent=0]
+include::../../../../samples/java-application-with-tests/pom.xml[tag=native-plugin]
+
+Finally, you will need to execute tests using the `integration-test` phase instead of `test`:
+
+```bash
+./mvn -Pnative integration-test
+```
+
+Please refer to the https://maven.apache.org/plugins/maven-shade-plugin[Maven Shade plugin documentation] for more details on how to configure shading and the https://maven.apache.org/plugins/maven-assembly-plugin[Maven Assembly plugin documentation] to tweak what to include in tests.
 
 == Javadocs
 

--- a/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationWithResourcesFunctionalTest.groovy
+++ b/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationWithResourcesFunctionalTest.groovy
@@ -53,6 +53,11 @@ class JavaApplicationWithResourcesFunctionalTest extends AbstractFunctionalTest 
         withSample("java-application-with-resources")
 
         buildFile << config
+        buildFile << """
+            nativeBuild {
+                verbose = true
+            }
+        """
 
         when:
         run 'nativeBuild'

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
@@ -47,10 +47,10 @@ import org.graalvm.buildtools.gradle.internal.AgentCommandLineProvider;
 import org.graalvm.buildtools.gradle.internal.GraalVMLogger;
 import org.graalvm.buildtools.gradle.internal.GradleUtils;
 import org.graalvm.buildtools.gradle.internal.ProcessGeneratedGraalResourceFiles;
-import org.graalvm.buildtools.utils.SharedConstants;
 import org.graalvm.buildtools.gradle.tasks.BuildNativeImageTask;
 import org.graalvm.buildtools.gradle.tasks.GenerateResourcesConfigFile;
 import org.graalvm.buildtools.gradle.tasks.NativeRunTask;
+import org.graalvm.buildtools.utils.SharedConstants;
 import org.gradle.api.Action;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
@@ -59,6 +59,7 @@ import org.gradle.api.file.ConfigurableFileCollection;
 import org.gradle.api.file.Directory;
 import org.gradle.api.file.DirectoryProperty;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.FileSystemLocation;
 import org.gradle.api.plugins.ApplicationPlugin;
 import org.gradle.api.plugins.JavaApplication;
 import org.gradle.api.plugins.JavaPlugin;
@@ -69,6 +70,8 @@ import org.gradle.api.tasks.SourceSet;
 import org.gradle.api.tasks.TaskCollection;
 import org.gradle.api.tasks.TaskContainer;
 import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.bundling.AbstractArchiveTask;
+import org.gradle.api.tasks.bundling.Jar;
 import org.gradle.api.tasks.testing.Test;
 import org.gradle.language.base.plugins.LifecycleBasePlugin;
 import org.gradle.process.JavaForkOptions;
@@ -76,8 +79,10 @@ import org.gradle.util.GFileUtils;
 
 import java.io.File;
 import java.util.Arrays;
+import java.util.Locale;
 import java.util.concurrent.Callable;
 import java.util.function.Consumer;
+import java.util.stream.Collectors;
 
 import static org.graalvm.buildtools.gradle.internal.GradleUtils.findConfiguration;
 import static org.graalvm.buildtools.gradle.internal.GradleUtils.findMainArtifacts;
@@ -148,6 +153,7 @@ public class NativeImagePlugin implements Plugin<Project> {
             task.getImage().convention(imageBuilder.map(t -> t.getOutputFile().get()));
             task.getRuntimeArgs().convention(buildExtension.getRuntimeArgs());
         });
+        configureClasspathJarFor(tasks, buildExtension, imageBuilder);
 
         TaskProvider<ProcessGeneratedGraalResourceFiles> processAgentFiles = registerProcessAgentFilesTask(project, PROCESS_AGENT_RESOURCES_TASK_NAME);
 
@@ -217,12 +223,32 @@ public class NativeImagePlugin implements Plugin<Project> {
             testExtension.getClasspath().from(testList);
             task.getAgentEnabled().set(testAgent);
         });
+        configureClasspathJarFor(tasks, testExtension, testImageBuilder);
 
         tasks.register(NATIVE_TEST_TASK_NAME, NativeRunTask.class, task -> {
             task.setDescription("Runs native-image compiled tests.");
             task.setGroup(LifecycleBasePlugin.VERIFICATION_GROUP);
             task.getImage().convention(testImageBuilder.map(t -> t.getOutputFile().get()));
             task.getRuntimeArgs().convention(testExtension.getRuntimeArgs());
+        });
+    }
+
+    private void configureClasspathJarFor(TaskContainer tasks, NativeImageOptions options, TaskProvider<BuildNativeImageTask> imageBuilder) {
+        String baseName = imageBuilder.getName();
+        TaskProvider<Jar> classpathJar = tasks.register(baseName + "ClasspathJar", Jar.class, jar -> {
+            jar.manifest(mn -> mn.getAttributes().put("Class-Path", options.getClasspath()
+                    .getElements()
+                    .map(cp -> cp.stream()
+                            .map(FileSystemLocation::getAsFile)
+                            .map(File::getAbsolutePath)
+                            .collect(Collectors.joining(" "))
+                    )));
+            jar.getArchiveBaseName().set(baseName.toLowerCase(Locale.ENGLISH) + "-classpath");
+        });
+        imageBuilder.configure(nit -> {
+            if (SharedConstants.IS_WINDOWS) {
+                nit.getClasspathJar().set(classpathJar.flatMap(AbstractArchiveTask::getArchiveFile));
+            }
         });
     }
 

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
@@ -254,7 +254,7 @@ public class NativeImagePlugin implements Plugin<Project> {
             jar.getArchiveBaseName().set(baseName.toLowerCase(Locale.ENGLISH) + "-classpath");
         });
         imageBuilder.configure(nit -> {
-            if (SharedConstants.IS_WINDOWS) {
+            if (options.getUseFatJar().get()) {
                 nit.getClasspathJar().set(classpathJar.flatMap(AbstractArchiveTask::getArchiveFile));
             }
         });

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
@@ -55,11 +55,12 @@ import org.gradle.api.Action;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
+import org.gradle.api.file.ArchiveOperations;
 import org.gradle.api.file.ConfigurableFileCollection;
 import org.gradle.api.file.Directory;
 import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.file.DuplicatesStrategy;
 import org.gradle.api.file.FileCollection;
-import org.gradle.api.file.FileSystemLocation;
 import org.gradle.api.plugins.ApplicationPlugin;
 import org.gradle.api.plugins.JavaApplication;
 import org.gradle.api.plugins.JavaPlugin;
@@ -77,6 +78,7 @@ import org.gradle.language.base.plugins.LifecycleBasePlugin;
 import org.gradle.process.JavaForkOptions;
 import org.gradle.util.GFileUtils;
 
+import javax.inject.Inject;
 import java.io.File;
 import java.util.Arrays;
 import java.util.Locale;
@@ -116,6 +118,11 @@ public class NativeImagePlugin implements Plugin<Project> {
     private static final String JUNIT_PLATFORM_LISTENERS_UID_TRACKING_OUTPUT_DIR = "junit.platform.listeners.uid.tracking.output.dir";
 
     private GraalVMLogger logger;
+
+    @Inject
+    public ArchiveOperations getArchiveOperations() {
+        throw new UnsupportedOperationException();
+    }
 
     @Override
     public void apply(Project project) {
@@ -236,13 +243,14 @@ public class NativeImagePlugin implements Plugin<Project> {
     private void configureClasspathJarFor(TaskContainer tasks, NativeImageOptions options, TaskProvider<BuildNativeImageTask> imageBuilder) {
         String baseName = imageBuilder.getName();
         TaskProvider<Jar> classpathJar = tasks.register(baseName + "ClasspathJar", Jar.class, jar -> {
-            jar.manifest(mn -> mn.getAttributes().put("Class-Path", options.getClasspath()
-                    .getElements()
-                    .map(cp -> cp.stream()
-                            .map(FileSystemLocation::getAsFile)
-                            .map(File::getAbsolutePath)
-                            .collect(Collectors.joining(" "))
-                    )));
+            jar.from(
+                    options.getClasspath()
+                            .getElements()
+                            .map(elems -> elems.stream()
+                                    .map(e -> getArchiveOperations().zipTree(e))
+                                    .collect(Collectors.toList()))
+            );
+            jar.setDuplicatesStrategy(DuplicatesStrategy.WARN);
             jar.getArchiveBaseName().set(baseName.toLowerCase(Locale.ENGLISH) + "-classpath");
         });
         imageBuilder.configure(nit -> {

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/NativeImageOptions.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/NativeImageOptions.java
@@ -42,6 +42,7 @@
 package org.graalvm.buildtools.gradle.dsl;
 
 import org.graalvm.buildtools.gradle.internal.GradleUtils;
+import org.graalvm.buildtools.utils.SharedConstants;
 import org.gradle.api.Action;
 import org.gradle.api.JavaVersion;
 import org.gradle.api.Project;
@@ -210,6 +211,7 @@ public abstract class NativeImageOptions {
         getAgent().convention(false);
         getSharedLibrary().convention(false);
         getImageName().convention(defaultImageName);
+        getUseFatJar().convention(SharedConstants.IS_WINDOWS);
         getJavaLauncher().convention(
                 toolchains.launcherFor(spec -> {
                     spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
@@ -355,4 +357,14 @@ public abstract class NativeImageOptions {
         return this;
     }
 
+    /**
+     * If set to true, this will build a fat jar of the image classpath
+     * instead of passing each jar individually to the classpath. This
+     * option can be used in case the classpath is too long and that
+     * invoking native image fails, which can happen on Windows.
+     *
+     * @return true if a fatjar should be used. Defaults to true for Windows, and false otherwise.
+     */
+    @Input
+    public abstract Property<Boolean> getUseFatJar();
 }

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/NativeImageCommandLineProvider.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/NativeImageCommandLineProvider.java
@@ -44,8 +44,10 @@ package org.graalvm.buildtools.gradle.internal;
 import org.graalvm.buildtools.gradle.dsl.NativeImageOptions;
 import org.gradle.api.Transformer;
 import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.file.RegularFile;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
 import org.gradle.api.tasks.Nested;
 import org.gradle.process.CommandLineArgumentProvider;
 
@@ -62,15 +64,18 @@ public class NativeImageCommandLineProvider implements CommandLineArgumentProvid
     private final Provider<Boolean> agentEnabled;
     private final Provider<String> executableName;
     private final Provider<String> outputDirectory;
+    private final Provider<RegularFile> classpathJar;
 
     public NativeImageCommandLineProvider(Provider<NativeImageOptions> options,
                                           Provider<Boolean> agentEnabled,
                                           Provider<String> executableName,
-                                          Provider<String> outputDirectory) {
+                                          Provider<String> outputDirectory,
+                                          Provider<RegularFile> classpathJar) {
         this.options = options;
         this.agentEnabled = agentEnabled;
         this.executableName = executableName;
         this.outputDirectory = outputDirectory;
+        this.classpathJar = classpathJar;
     }
 
     @Nested
@@ -93,13 +98,22 @@ public class NativeImageCommandLineProvider implements CommandLineArgumentProvid
         return outputDirectory;
     }
 
+    @InputFile
+    public Provider<RegularFile> getClasspathJar() {
+        return classpathJar;
+    }
+
     @Override
     public List<String> asArguments() {
         NativeImageOptions options = getOptions().get();
         List<String> cliArgs = new ArrayList<>(20);
 
         cliArgs.add("-cp");
-        cliArgs.add(options.getClasspath().getAsPath());
+        if (classpathJar.isPresent()) {
+            cliArgs.add(classpathJar.get().getAsFile().getAbsolutePath());
+        } else {
+            cliArgs.add(options.getClasspath().getAsPath());
+        }
 
         appendBooleanOption(cliArgs, options.getDebug(), "-H:GenerateDebugInfo=1");
         appendBooleanOption(cliArgs, options.getFallback().map(NEGATE), "--no-fallback");

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/BuildNativeImageTask.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/BuildNativeImageTask.java
@@ -49,10 +49,12 @@ import org.gradle.api.GradleException;
 import org.gradle.api.file.Directory;
 import org.gradle.api.file.DirectoryProperty;
 import org.gradle.api.file.RegularFile;
+import org.gradle.api.file.RegularFileProperty;
 import org.gradle.api.provider.Property;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.provider.ProviderFactory;
 import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
 import org.gradle.api.tasks.Internal;
 import org.gradle.api.tasks.Nested;
 import org.gradle.api.tasks.Optional;
@@ -112,6 +114,10 @@ public abstract class BuildNativeImageTask extends DefaultTask {
     @Inject
     protected abstract ProviderFactory getProviders();
 
+    @InputFile
+    @Optional
+    public abstract RegularFileProperty getClasspathJar();
+
     public BuildNativeImageTask() {
         DirectoryProperty buildDir = getProject().getLayout().getBuildDirectory();
         Provider<Directory> outputDir = buildDir.dir("native/" + getName());
@@ -132,8 +138,8 @@ public abstract class BuildNativeImageTask extends DefaultTask {
                 getExecutableName(),
                 // Can't use getOutputDirectory().map(...) because Gradle would complain that we use
                 // a mapped value before the task was called, when we are actually calling it...
-                getProviders().provider(() -> getOutputDirectory().getAsFile().get().getAbsolutePath())
-        ).asArguments();
+                getProviders().provider(() -> getOutputDirectory().getAsFile().get().getAbsolutePath()),
+                getClasspathJar()).asArguments();
     }
 
     // This property provides access to the service instance

--- a/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationFunctionalTest.groovy
+++ b/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationFunctionalTest.groovy
@@ -52,4 +52,15 @@ class JavaApplicationFunctionalTest extends AbstractGraalVMMavenFunctionalTest {
         buildSucceeded
         outputContains "Hello, native!"
     }
+
+    def "can build and execute a native image with the Maven plugin and the shade plugin"() {
+        withSample("java-application")
+
+        when:
+        mvn '-Pshaded', '-DskipTests', 'package', 'exec:exec@native'
+
+        then:
+        buildSucceeded
+        outputContains "Hello, native!"
+    }
 }

--- a/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithTestsFunctionalTest.groovy
+++ b/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithTestsFunctionalTest.groovy
@@ -66,4 +66,29 @@ class JavaApplicationWithTestsFunctionalTest extends AbstractGraalVMMavenFunctio
 [         0 tests failed          ]
 """.trim()
     }
+
+    def "can run tests in a native image with the Maven plugin using shading"() {
+        withSample("java-application-with-tests")
+
+        when:
+        mvn '-Pshaded', 'integration-test'
+
+        then:
+        buildSucceeded
+        outputContains "[junit-platform-native] Running in 'test listener' mode"
+        outputContains """
+[         3 containers found      ]
+[         0 containers skipped    ]
+[         3 containers started    ]
+[         0 containers aborted    ]
+[         3 containers successful ]
+[         0 containers failed     ]
+[         6 tests found           ]
+[         0 tests skipped         ]
+[         6 tests started         ]
+[         0 tests aborted         ]
+[         6 tests successful      ]
+[         0 tests failed          ]
+""".trim()
+    }
 }

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeBuildMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeBuildMojo.java
@@ -63,6 +63,7 @@ import java.nio.file.FileSystem;
 import java.nio.file.FileSystems;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
@@ -94,7 +95,10 @@ public class NativeBuildMojo extends AbstractNativeMojo {
     @Parameter(defaultValue = "${mojoExecution}")
     private MojoExecution mojoExecution;
 
-    private final List<Path> classpath = new ArrayList<>();
+    @Parameter(property = "classpath")
+    private List<String> classpath;
+
+    private final List<Path> imageClasspath = new ArrayList<>();
 
     private PluginParameterExpressionEvaluator evaluator;
 
@@ -105,14 +109,18 @@ public class NativeBuildMojo extends AbstractNativeMojo {
         }
         evaluator = new PluginParameterExpressionEvaluator(session, mojoExecution);
 
-        classpath.clear();
-        List<String> imageClasspathScopes = Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_RUNTIME);
-        project.setArtifactFilter(artifact -> imageClasspathScopes.contains(artifact.getScope()));
-        for (Artifact dependency : project.getArtifacts()) {
-            addClasspath(dependency);
+        imageClasspath.clear();
+        if (classpath != null && !classpath.isEmpty()) {
+            imageClasspath.addAll(classpath.stream().map(Paths::get).collect(Collectors.toList()));
+        } else {
+            List<String> imageClasspathScopes = Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_RUNTIME);
+            project.setArtifactFilter(artifact -> imageClasspathScopes.contains(artifact.getScope()));
+            for (Artifact dependency : project.getArtifacts()) {
+                addClasspath(dependency);
+            }
+            addClasspath(project.getArtifact());
         }
-        addClasspath(project.getArtifact());
-        String classpathStr = classpath.stream().map(Path::toString).collect(Collectors.joining(File.pathSeparator));
+        String classpathStr = imageClasspath.stream().map(Path::toString).collect(Collectors.joining(File.pathSeparator));
 
         Path nativeImageExecutable = Utils.getNativeImage();
 
@@ -170,7 +178,7 @@ public class NativeBuildMojo extends AbstractNativeMojo {
             throw new MojoExecutionException("Artifact " + artifact + "cannot be added to image classpath", e);
         }
 
-        classpath.add(jarFilePath);
+        imageClasspath.add(jarFilePath);
     }
 
     private Path getWorkingDirectory() {

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeTestMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeTestMojo.java
@@ -79,6 +79,9 @@ public class NativeTestMojo extends AbstractNativeMojo {
     @Parameter(property = "skipTests", defaultValue = "false")
     private boolean skipTests;
 
+    @Parameter(property = "classpath")
+    private List<String> classpath;
+
     @Override
     public void execute() throws MojoExecutionException, MojoFailureException {
         if (skipTests) {
@@ -182,6 +185,9 @@ public class NativeTestMojo extends AbstractNativeMojo {
     }
 
     private String getClassPath() throws MojoFailureException {
+        if (classpath != null && !classpath.isEmpty()) {
+            return String.join(File.pathSeparator, classpath);
+        }
         try {
             List<Artifact> pluginDependencies = pluginArtifacts.stream()
                     .filter(it -> it.getGroupId().startsWith(Utils.MAVEN_GROUP_ID) || it.getGroupId().startsWith("org.junit"))

--- a/samples/java-application-with-tests/pom.xml
+++ b/samples/java-application-with-tests/pom.xml
@@ -109,10 +109,97 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>shaded</id>
+            <build>
+                <plugins>
+                    <!-- tag::shade-plugin[] -->
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-shade-plugin</artifactId>
+                        <version>3.2.4</version>
+                        <executions>
+                            <execution>
+                                <phase>package</phase>
+                                <goals>
+                                    <goal>shade</goal>
+                                </goals>
+                            </execution>
+                        </executions>
+                        <configuration>
+                            <shadedArtifactAttached>true</shadedArtifactAttached>
+                        </configuration>
+                    </plugin>
+                    <!-- end::shade-plugin[] -->
+
+                    <!-- tag::assembly-plugin[] -->
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-assembly-plugin</artifactId>
+                        <version>2.4.1</version>
+                        <configuration>
+                            <descriptors>
+                                <descriptor>src/assembly/test-jar-with-dependencies.xml</descriptor>
+                            </descriptors>
+                        </configuration>
+                        <executions>
+                            <execution>
+                                <id>make-test-jar</id>
+                                <phase>package</phase>
+                                <goals>
+                                    <goal>single</goal>
+                                </goals>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <!-- end::assembly-plugin[] -->
+
+                    <!-- tag::native-plugin[] -->
+                    <plugin>
+                        <groupId>org.graalvm.buildtools</groupId>
+                        <artifactId>native-maven-plugin</artifactId>
+                        <version>${native.maven.plugin.version}</version>
+                        <executions>
+                            <execution>
+                                <id>test-native</id>
+                                <goals>
+                                    <goal>test</goal>
+                                </goals>
+                                <phase>integration-test</phase>
+                                <configuration>
+                                    <classpath>
+                                        <param>${project.build.directory}/${project.artifactId}-${project.version}-tests.jar</param>
+                                    </classpath>
+                                </configuration>
+                            </execution>
+                            <execution>
+                                <id>build-native</id>
+                                <goals>
+                                    <goal>build</goal>
+                                </goals>
+                                <phase>package</phase>
+                                <configuration>
+                                    <classpath>
+                                        <param>${project.build.directory}/${project.artifactId}-${project.version}-shaded.jar</param>
+                                    </classpath>
+                                </configuration>
+                            </execution>
+                        </executions>
+                        <configuration>
+                            <skip>false</skip>
+                            <imageName>${imageName}</imageName>
+                            <buildArgs>
+                                <buildArg>--no-fallback</buildArg>
+                            </buildArgs>
+                        </configuration>
+                    </plugin>
+                    <!-- end::native-plugin[] -->
+                </plugins>
+            </build>
+        </profile>
     </profiles>
 
     <build>
-        <finalName>${project.artifactId}</finalName>
         <plugins>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>

--- a/samples/java-application-with-tests/src/assembly/test-jar-with-dependencies.xml
+++ b/samples/java-application-with-tests/src/assembly/test-jar-with-dependencies.xml
@@ -1,0 +1,71 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2020, 2021, Oracle and/or its affiliates. All rights reserved.
+    DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+
+    The Universal Permissive License (UPL), Version 1.0
+
+    Subject to the condition set forth below, permission is hereby granted to any
+    person obtaining a copy of this software, associated documentation and/or
+    data (collectively the "Software"), free of charge and under any and all
+    copyright rights in the Software, and any and all patent rights owned or
+    freely licensable by each licensor hereunder covering either (i) the
+    unmodified Software as contributed to or provided by such licensor, or (ii)
+    the Larger Works (as defined below), to deal in both
+
+    (a) the Software, and
+
+    (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+    one is included with the Software each a "Larger Work" to which the Software
+    is contributed by such licensors),
+
+    without restriction, including without limitation the rights to copy, create
+    derivative works of, display, perform, and distribute the Software and make,
+    use, sell, offer for sale, import, export, have made, and have sold the
+    Software and the Larger Work(s), and to sublicense the foregoing rights on
+    either these or other terms.
+
+    This license is subject to the following condition:
+
+    The above copyright notice and either this complete permission notice or at a
+    minimum a reference to the UPL must be included in all copies or substantial
+    portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+-->
+
+<!-- tag::assembly[] -->
+<assembly xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2"
+          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 http://maven.apache.org/xsd/assembly-1.1.2.xsd">
+    <id>tests</id>
+    <formats>
+        <format>jar</format>
+    </formats>
+    <fileSets>
+        <fileSet>
+            <directory>${project.build.directory}/test-classes</directory>
+            <outputDirectory>/</outputDirectory>
+        </fileSet>
+        <fileSet>
+            <directory>${project.build.outputDirectory}</directory>
+            <outputDirectory>/</outputDirectory>
+        </fileSet>
+    </fileSets>
+    <includeBaseDirectory>false</includeBaseDirectory>
+    <dependencySets>
+        <dependencySet>
+            <outputDirectory>/</outputDirectory>
+            <useProjectArtifact>true</useProjectArtifact>
+            <unpack>true</unpack>
+            <scope>test</scope>
+        </dependencySet>
+    </dependencySets>
+</assembly>
+<!-- end::assembly[] -->

--- a/samples/java-application/pom.xml
+++ b/samples/java-application/pom.xml
@@ -95,6 +95,58 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>shaded</id>
+            <build>
+                <plugins>
+                    <!-- tag::shade-plugin[] -->
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-shade-plugin</artifactId>
+                        <version>3.2.4</version>
+                        <executions>
+                            <execution>
+                                <phase>package</phase>
+                                <goals>
+                                    <goal>shade</goal>
+                                </goals>
+                                <configuration>
+                                    <shadedArtifactAttached>true</shadedArtifactAttached>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <!-- end::shade-plugin[] -->
+
+                    <!-- tag::native-plugin[] -->
+                    <plugin>
+                        <groupId>org.graalvm.buildtools</groupId>
+                        <artifactId>native-maven-plugin</artifactId>
+                        <version>${native.maven.plugin.version}</version>
+                        <executions>
+                            <execution>
+                                <id>build-native</id>
+                                <goals>
+                                    <goal>build</goal>
+                                </goals>
+                                <phase>package</phase>
+                            </execution>
+                        </executions>
+                        <configuration>
+                            <skip>false</skip>
+                            <imageName>${imageName}</imageName>
+                            <buildArgs>
+                                <buildArg>--no-fallback</buildArg>
+                            </buildArgs>
+                            <classpath>
+                                <param>${project.build.directory}/${project.artifactId}-${project.version}-shaded.jar</param>
+                            </classpath>
+                        </configuration>
+                    </plugin>
+                    <!-- end::native-plugin[] -->
+                </plugins>
+            </build>
+        </profile>
     </profiles>
 
     <build>

